### PR TITLE
built a component for delete button in tables

### LINKED ISSUE #9863: [UX]: Make lists/table design & table actions and make confirm delete action smoother
image

image

image

Make the design a bit more cohesive and make the confirm delete action more inline like for github labels (instead of the current modal).

image

### PR DESCRIPTION
**Pull request by Ijin08**
instead of using a modal to confirm it now does it in the row of the table, created a sass file for the component, the component uses css transitions for animation

fixes #9863 

delete-button-before
delete-button-after

### CODE REVIEW

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import React, { Component } from 'react';
+
+export default class DeleteButton extends Component<any, any> {
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: - This can be made a `PureComponent` because its rendering depends entirely on state and props.
- It's easy to add `DeleteButtonProps` and `DeleteButtonState` types here, so we should do it, see [this guide](https://github.com/piotrwitek/react-redux-typescript-guide#stateful-components---class)

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  state = {
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: It's a bit unusual to use render-specifics like CSS class names as the state values. I would recommend against it, because it's much easier to reason later in the render function on what's going on. Try modelling the state with true/false values instead.

The state is essentially binary: delete vs showing the confirm button, so one state value (e.g. `confirmed: false`) could suffice.

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    if (event) {
+      event.preventDefault();
+    }
+
+    this.setState({
+      deleteButton: 'delete-button--hide',
+    });
+
+    setTimeout(() => {
+      this.setState({
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: You could probably remove the `--removed` property, and always display the confirm button invisibly and simply have a `onClickConfirm` handler that checks if `showConfirm` is true, and only then execute the `confirmDelete` prop.

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        deleteButton: 'delete-button--removed',
+      });
+    }, 100);
+
+    setTimeout(() => {
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: Could you use CSS delays for transitions to get rid of the all `setTimeout`?

File: public/sass/components/_delete_button.scss [hunk 5]
```diff
@@ -0,0 +1,49 @@
+.delete-button-container {
+  max-width: 24px;
+  width: 24px;
+  direction: rtl;
```
Comment (davkal) [inline public/sass/components/_delete_button.scss]: Why is this needed?

**Review by davkal — CHANGES_REQUESTED** [review]
This interaction is great! 

Blocker:
- This needs types (or tests).

Minor:
- I think it's worth trying to model the component state with booleans, and then use the CSS class names in the render method, this will make it much clearer what's going on. (Let me know if you need help with this.)
- Also worth looking into getting rid of the `setTimeout` in favor of CSS mechanics.
- I did manage to double-click to instantly confirm

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import React, { Component } from 'react';
+
+export default class DeleteButton extends Component<any, any> {
+  state = {
+    deleteButton: 'delete-button--show',
+    confirmSpan: 'confirm-delete--removed',
+  };
+
+  handleDelete = event => {
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: Should be named `onClickDelete`

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }, 100);
+
+    setTimeout(() => {
+      this.setState({
+        confirmSpan: 'confirm-delete--show',
+      });
+    }, 150);
+  };
+
+  cancelDelete = event => {
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: Should be named `onClickCancel`

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      <span className="delete-button-container">
+        <a className={this.state.deleteButton + ' btn btn-danger btn-small'} onClick={this.handleDelete}>
+          <i className="fa fa-remove" />
+        </a>
+        <span className="confirm-delete-container">
+          <span className={this.state.confirmSpan}>
+            <a className="btn btn-small" onClick={this.cancelDelete}>
+              Cancel
+            </a>
+            <a className="btn btn-danger btn-small" onClick={confirmDelete}>
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: I'd wrap this in a local handler: `onClickConfirm` that calls `this.props.onConfirmDelete`

File: public/app/core/components/DeleteButton/DeleteButton.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    setTimeout(() => {
+      this.setState({
+        deleteButton: 'delete-button--show',
+      });
+    }, 190);
+  };
+
+  render() {
+    const { confirmDelete } = this.props;
```
Comment (davkal) [inline public/app/core/components/DeleteButton/DeleteButton.tsx]: Should be named `onConfirmDelete`

File: public/sass/components/_delete_button.scss (hunk 5 above)
Comment (Ijin08) [reply public/sass/components/_delete_button.scss]: This is needed so that when you push the delete button it doesn't impact the rest of the row. So when the confirm is shown it is overflowing this div and that is why reverse direction is needed.

Comment (davkal) [reply public/sass/components/_delete_button.scss]: Ah, ok. Could do with a comment then.

**Review by davkal — APPROVED** [review]

File: public/app/containers/Teams/TeamList.tsx [hunk 10]
```diff
@@ -102,7 +88,7 @@ export class TeamList extends React.Component<Props, any> {
             </a>
           </div>
 
-          <div className="admin-list-table">
+          <div className="admin-list-table tr-overflow">
```
[the remark this answers is not in the document]
Comment (Ijin08) [reply public/app/containers/Teams/TeamList.tsx]: I think that is an leftover from an earlier solution, will look into it!

[the remark this answers is not in the document]
Comment (Ijin08) [reply public/app/containers/Teams/TeamList.tsx]: Yeah, that shouldn't be there. Good catch. Pushed a fix.
